### ISSUE TITLE
LogSegment: Index file

### ISSUE THREAD

**Issue by tferdous17** [tferdous17/flux#21]
In Kafka, the `.index` file plays a crucial role in efficiently locating records within a log segment. Without this file, Kafka would have to scan through each individual record in the data file until it finds what it's looking for, which is **slow**.

The purpose of the `.index` file is to map offsets (unique identifiers for records in Kafka) to their corresponding physical byte locations within the data file (`.log`). This allows Kafka, and our system, to quickly locate records by their offset which is important.

**Functional requirements:**
- Implement an `.index` file within `LogSegment`
- Figure out a way to map each record's offset to that record's _byte position_ within the `.log` file. 
    - For example, a simplified and unserialized entry example might look like this: `Offset 19 --> Byte Position 8371`
- If possible, implement a binary search algorithm to find the correct byte location of a given offset in O(log N) time

Tl;dr: The **log segment file** holds the actual records, and the **index file** holds the mapping from offsets to byte positions within the log segment file.
